packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt

```diff
@@ -0,0 +1,143 @@
+package expo.modules.kotlin.activityresult
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.os.Bundle
+import android.os.Parcel
+import android.util.Base64
+import androidx.core.os.bundleOf
+import java.io.Serializable
+import java.util.*
+import kotlin.collections.ArrayList
+
+const val EXPIRE_KEY = "expire"
+const val EXPIRATION_TIME = 5 * 60 * 1000 // 5 min
+
+/**
+ * This class serves as a persistable store that accepts different kinds of data that have to
+ * be preserved between Activity destruction and recreation.
+ * For each kind of data there's a separate pair of methods for storing and retrieving the data.
+ *
+ * Ideally we would use [android.app.Activity.onSaveInstanceState] and [android.app.Activity.onCreate]
+ * alongside with `savedBundleState`, but it's blocked by https://github.com/software-mansion/react-native-screens/issues/17#issuecomment-424704067
+ */
+class DataPersistor(context: Context) {
+  private val sharedPreferences: SharedPreferences = context.getSharedPreferences("expo.modules.kotlin.PersistentDataManager", Context.MODE_PRIVATE)
+
+  private val accumulator = Bundle()
+
+  @Suppress("PropertyName")
+  private var _retrievedData: Bundle? = null
+  private val retrievedData: Bundle
+    get() = _retrievedData ?: retrieveData().also { _retrievedData = it }
```

Comment:
```suggestion
  private val retrievedData by lazy { retrieveData() }
```
It should work in the same way ;)